core/trino-main/src/main/java/io/trino/execution/executor2/scheduler/SchedulingQueue.java

```diff
@@ -0,0 +1,531 @@
+
+        T task = info.dequeue(expectedWeight);
+        verify(task != null);
+
+        doTransition(group, info);
+
+        return task;
+    }
+
+    public T peek()
+    {
+        G group = runnableQueue.peek();
+
+        if (group == null) {
+            return null;
+        }
+
+        Group<T> info = groups.get(group);
+        verify(info.state() == RUNNABLE, "Group is not runnable: %s", group);
+
+        T task = info.peek();
+        checkState(task != null);
+
+        return task;
+    }
+
+    private void doTransition(G group, Group<T> info)
+    {
+        switch (info.state()) {
+            case RUNNABLE -> transitionToRunnable(group, info);
+            case RUNNING -> transitionToRunning(group, info);
+            case BLOCKED -> transitionToBlocked(group, info);
+        }
+
+        verifyState(group);
+    }
+
+    private void transitionToRunning(G group, Group<T> info)
+    {
+        checkArgument(info.state() == RUNNING);
+
+        baselineWeights.addOrReplace(group, info.weight());
+    }
+
+    private void transitionToBlocked(G group, Group<T> info)
+    {
+        checkArgument(info.state() == BLOCKED);
+
+        blocked.put(group, new Blocked(info.weight() - baselineWeight()));
+        baselineWeights.removeIfPresent(group);
+        runnableQueue.removeIfPresent(group);
+    }
+
+    private void transitionToRunnable(G group, Group<T> info)
+    {
+        checkArgument(info.state() == RUNNABLE);
+
+        runnableQueue.addOrReplace(group, info.weight());
+        baselineWeights.addOrReplace(group, info.weight());
+    }
+
+    public State state(G group)
+    {
+        Group<T> info = groups.get(group);
+        checkArgument(info != null, "Unknown group: %s", group);
+
+        return info.state();
+    }
+
+    private long baselineWeight()
+    {
+        if (baselineWeights.isEmpty()) {
+            return 0;
+        }
+
+        return baselineWeights.nextPriority();
+    }
+
+    private void verifyState(G groupKey)
+    {
+        Group<T> group = groups.get(groupKey);
+        checkArgument(group != null, "Unknown group: %s", groupKey);
+
+        switch (group.state()) {
+            case BLOCKED -> {
+                checkState(!runnableQueue.contains(groupKey), "Group in BLOCKED state should not be in queue: %s", groupKey);
+                checkState(!baselineWeights.contains(groupKey));
+            }
+            case RUNNABLE -> {
+                checkState(runnableQueue.contains(groupKey), "Group in RUNNABLE state should be in queue: %s", groupKey);
+                checkState(baselineWeights.contains(groupKey));
+            }
+            case RUNNING -> {
+                checkState(!runnableQueue.contains(groupKey), "Group in RUNNING state should not be in queue: %s", groupKey);
+                checkState(baselineWeights.contains(groupKey));
+            }
+        }
+    }
+
+    @Override
+    public String toString()
```

Comment:
Do you really want this in `toString` or do you want an "info" method?

Reply:
I could move it to an info method, but It comes in very handing when running in the debugger.